### DISCUSSION TITLE
Add constructors and named exports

### DISCUSSION THREAD

**Discussion by dgregoire** [ahmedsakr/wstrade-api#92] [Ideas]
In the cases where you would want to develop a server app around your lib - currently there's one global instance of the data (tokens, ...) that all your modules refer to when pulling data from WS.

Would you consider having constructors and a way to have multiple users logged in and using your module concurrently?

**Comment by ahmedsakr** [comment]
The current design definitely does not support concurrency. I'll admit that's likely the biggest flaw in `v1` at this moment. When I was designing v1 last year, I was under the impression that the users of this library were retail investors that wanted programmatic access to their WST account. In hindsight, I realize that this was a major oversight. As a matter of fact, I believe `v0.x.x` did support concurrent usage (as you had to provide the tokens manually to every API call). So this regression is something that needs to be addressed and I am happy your brought it up.

I am very interested in supporting concurrent usage. I am struggling to brainstorm a workflow that would fit into the current version while being backwards compatible. I'll have to get back to this soon.

**Reply by ahmedsakr** [reply]
@dgregoire I'd like your opinion on this change that I am proposing to support concurrent usage of this library:

The library would export a `Client` class that would create objects that contain the entire library APIs, something along the following:

```javascript

import * as trade from 'wstrade-api';

const user1 = new trade.Client();
const user2 = new trade.Client();
// as many users as you want...

// you can then use the library on every object as usual
user1.auth.login(...)
```

To make it backwards compatible, I'll export a default `Client` object in replacement of the constant object that is being exported now.

**Reply by dgregoire** [reply]
This looks good to me! I was thinking along the same line, session or context.

**Reply by ahmedsakr** [reply]
Fantastic. Alright then, I'll raise an issue and prioritize this over other outstanding issues. Thanks for bringing it up.

**Reply by ahmedsakr** [reply]
@dgregoire I have put up a PR for supporting concurrency. I'd appreciate it if you can weigh in on it: https://github.com/ahmedsakr/wstrade-api/pull/102

**Reply by ahmedsakr** [reply]
@dgregoire the important part is https://github.com/ahmedsakr/wstrade-api/pull/102/files#diff-bfe9874d239014961b1ae4e89875a6155667db834a410aaaa2ebe3cf89820556. This is where I implemented the session constructor. Moreover, you will notice that the `headers` and `config` modules remain mutual between all sessions for consistency.

**Reply by dgregoire** [reply]
@ahmedsakr it looks good to me. You've taken care of the `on` handler for the `2fa` problem in one fell swoop as well as far as I can tell - I'm going to start testing with it and report back if I stumble unto anything.

good work there 👍

**Reply by ahmedsakr** [reply]
Which `2fa` problem? do you mean this? https://github.com/ahmedsakr/wstrade-api/issues/101

I didn't change any handlers in this PR

**Reply by dgregoire** [reply]
I meant you have a `on` handler per session - I initially thought this would get done as part of #95 but now I realize there was no way around doing it now.

**Reply by ahmedsakr** [reply]
ahh yep yep, that's right. You'll have to register the `on` handlers for every session you create.

**Reply by dgregoire** [reply]
I have the branch up in production and I'll make a pass at #95, thank you for your work!

**Reply by ahmedsakr** [reply]
No problem. I tested the branch extensively and everything seems to work as intended. I merged it into main and it's available as of `1.8.0` on npm.

**Comment by ahmedsakr** [comment]
@dgregoire Hey Dominic. Sorry not related to this resolved issue, but I wanted to ask if you were interested in becoming the NPM maintainer or contributor for this project. I actually am retiring myself from this project because it's been a while (but also because i work at Wealthsimple now).
